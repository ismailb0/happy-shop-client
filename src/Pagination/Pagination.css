.pagination {
  flex-grow: 5;
  display: flex;
  padding: 0 3%;
  align-items: center;
  justify-content: flex-start;
}

.number {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  min-width: 8%;
  flex-grow: 1;
}

.selected {
  color: #ff008c;
}
