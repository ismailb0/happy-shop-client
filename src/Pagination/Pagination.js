import React, { Component, Fragment } from 'react';

import './Pagination.css';

class Pagination extends Component {

  state = {
    currentPage: this.props.page,
    numbersToDisplay: [],
    pageNumbers: []
  };

  componentWillReceiveProps(nextProps) {
    const pageNumbers = this.getPageNumbers(nextProps.totalNumberOfProducts, nextProps.resultsPerPage)
    this.updateNumbersToDisplay(pageNumbers, this.state.currentPage)
    if (nextProps.page !== this.props.page) {
      this.setState({
        currentPage: nextProps.page
      })
    }
  }

  handleClick = (value) => {
    this.props.handlePageChange(value)
    this.setState({ currentPage: value }, () => this.updateNumbersToDisplay(
      this.state.pageNumbers,
      value
    ));
  }

  getNumber = (number) => {
    return (
      <div
        className={(number === this.state.currentPage) ? 'number selected' : 'number'}
        key={number}
        id={number}
        onClick={(event) => this.handleClick(parseInt(event.target.id))}
      >
        {number}
      </div>
    );
  }

  getNumberWithEllipsis = (number) => {
    return (
      <Fragment key={number}>
        <div className='number'>...</div>
        <div
          className={(number === this.state.currentPage) ? 'number selected' : 'number'}
          key={number}
          id={number}
          onClick={(event) => this.handleClick(parseInt(event.target.id))}
        >
          {number}
        </div>
      </Fragment>
    );
  }

  updateNumbersToDisplay = (pageNumbers, currentPage) => {
    const firstPage = pageNumbers[0]
    const [ lastPage ] = pageNumbers.slice(-1)

    let numbersToDisplay = [firstPage]

    if (currentPage > firstPage && currentPage < lastPage) {
      numbersToDisplay.push(currentPage - 1)
      numbersToDisplay.push(currentPage)
      numbersToDisplay.push(currentPage + 1)
    } else if (currentPage === firstPage) {
      numbersToDisplay.push(currentPage)
      numbersToDisplay.push(currentPage + 1)
      numbersToDisplay.push(currentPage + 2)
    } else if (currentPage === lastPage) {
      numbersToDisplay.push(currentPage - 2)
      numbersToDisplay.push(currentPage - 1)
      numbersToDisplay.push(currentPage)
    }

    numbersToDisplay.push(lastPage)

    // Remove duplicates
    numbersToDisplay = numbersToDisplay.filter((value,index) => numbersToDisplay.indexOf(value) === index)

    // Remove out of bounds values
    numbersToDisplay = numbersToDisplay.filter((value, index) => (value <= lastPage && value >= firstPage))

    this.setState({ numbersToDisplay })
  }

  getPageNumbers = (totalNumberOfProducts, resultsPerPage) => {
    const pageNumbers = [];
    const numberOfPages = Math.ceil(totalNumberOfProducts / resultsPerPage)
    for (let i = 1; i <= numberOfPages; i++) {
      pageNumbers.push(i);
    }
    this.setState({ pageNumbers });
    return pageNumbers
  }

  render() {
    const { numbersToDisplay } = this.state

    return (
      <div className='pagination'>
        {
          numbersToDisplay.map((element, index) => {
            if (index > 0 && element !== numbersToDisplay[index - 1] + 1) {
              return this.getNumberWithEllipsis(element)
            } else {
              return this.getNumber(element)
            }
          })
        }
      </div>
    );
  }
}

export default Pagination;
