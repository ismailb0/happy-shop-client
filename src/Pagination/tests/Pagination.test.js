import React from 'react';
import { shallow } from "enzyme";
import Pagination from '../Pagination';

describe('<Pagination />', () => {

  const props = {
    handlePageChange: jest.fn(),
    totalNumberOfProducts: 48,
    resultsPerPage: 10
  }

  it('getPageNumbers should get all possible pagination numbers and save them in the state', () => {
    const renderedComponent = shallow(<Pagination {...props} />);
    const componentInstance = renderedComponent.instance()
    expect(componentInstance.getPageNumbers(48, 10)).toEqual([1, 2, 3, 4, 5])
    expect(componentInstance.state.pageNumbers).toEqual([1, 2, 3, 4, 5])
  });

  it('updateNumbersToDisplay should set numbers to display in pagination in the state', () => {
    const renderedComponent = shallow(<Pagination {...props} />);
    const componentInstance = renderedComponent.instance()
    componentInstance.updateNumbersToDisplay([1, 2, 3, 4, 5, 6], 1)
    expect(componentInstance.state.numbersToDisplay).toEqual([1, 2, 3, 6])
    componentInstance.updateNumbersToDisplay([1, 2, 3, 4, 5, 6, 8, 9, 10], 4)
    expect(componentInstance.state.numbersToDisplay).toEqual([1, 3, 4, 5, 10])
    componentInstance.updateNumbersToDisplay([1, 2, 3, 4, 5, 6, 8, 9, 10], 8)
    expect(componentInstance.state.numbersToDisplay).toEqual([1, 7, 8, 9, 10])
  });

  it('handleClick should call handlePageChange and update numbers to display', () => {
    const renderedComponent = shallow(<Pagination {...props} />);
    const componentInstance = renderedComponent.instance()
    jest.spyOn(componentInstance, 'updateNumbersToDisplay');
    componentInstance.handleClick(1)
    expect(props.handlePageChange).toHaveBeenCalledWith(1)
    expect(componentInstance.state.currentPage).toEqual(1)
    expect(componentInstance.updateNumbersToDisplay).toHaveBeenCalledWith([], 1)
  });

  it('should match snapshot with Ellipsis', () => {
    const renderedComponent = shallow(<Pagination {...props} />);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should match snapshot without Ellipsis', () => {
    const props = {
      handlePageChange: jest.fn(),
      totalNumberOfProducts: 25,
      resultsPerPage: 10
    }
    const renderedComponent = shallow(<Pagination {...props} />);
    expect(renderedComponent).toMatchSnapshot();
  });

});
