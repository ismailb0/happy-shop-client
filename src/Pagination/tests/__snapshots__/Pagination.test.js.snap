// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should match snapshot with Ellipsis 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    handlePageChange={
      [MockFunction] {
        "calls": Array [
          Array [
            1,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    page={1}
    resultsPerPage={10}
    totalNumberOfProducts={48}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "pagination",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "pagination",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<Pagination /> should match snapshot without Ellipsis 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    handlePageChange={[MockFunction]}
    resultsPerPage={10}
    totalNumberOfProducts={25}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "pagination",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "pagination",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
